### PR TITLE
Implement Test resource

### PR DESCRIPTION
**Pull request by Fenthick**
Issue: KNTK-245
Based on: #12

### CODE REVIEW

File: synthetics/test_conversions.go [hunk 1]
```diff
@@ -214,3 +217,343 @@ func userInfoToMapSlice(obj *synthetics.V202101beta1UserInfo) []map[string]inter
 		"full_name": obj.FullName,
 	}}
 }
+
+// resourceDataToTest converts TF resource data to Test object. Only user-writable attributes are set.
+func resourceDataToTest(d *schema.ResourceData) (*synthetics.V202101beta1Test, error) {
+	test := synthetics.NewV202101beta1Test()
+
+	if id, ok := d.GetOk(idKey); ok {
```
Comment (mmac-m3a) [inline synthetics/test_conversions.go:225]: Is this `if <var> := <something>; <var> { ... }` a common Golang idiom? It looks unnecessarily obtuse to me (but I'd not call myself a Go programmer). Why not keep it simple?

```
id, ok := d.GetOk(idKey)
if ok {
    test.SetId(id.(string))
}
```

File: synthetics/test_schema.go [hunk 2]
```diff
@@ -4,46 +4,53 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
+// TODO(dfurman): mark required attributes, when they are specified in the OpenAPI definitions
+// TODO(dfurman): provide descriptions, when they are specified in the OpenAPI definitions
+
 func makeTestSchema(mode schemaMode) map[string]*schema.Schema {
 	return map[string]*schema.Schema{
 		"id": {
```
Comment (mmac-m3a) [inline synthetics/test_schema.go:12]: Can/should this be `idKey` to keep things consistent?

File: synthetics/test_schema.go [hunk 3]
```diff
@@ -321,44 +372,50 @@ func makeTestTraceSchema(mode schemaMode) *schema.Schema {
...
 			Computed: computedOnRead(mode),
 		},
 	})
 }
 
-func makeUserInfoSchema(mode schemaMode) *schema.Schema {
-	return makeNestedObjectSchema(mode, map[string]*schema.Schema{
+func makeUserInfoSchema() *schema.Schema {
+	return makeReadOnlyNestedObjectSchema(map[string]*schema.Schema{
 		"id": {
```
Comment (mmac-m3a) [inline synthetics/test_schema.go:408]: `idKey`?

**Review by mmac-m3a — APPROVED** [review]
LGTM. Consider my comments for the next pull request.

File: synthetics/test_conversions.go (hunk 1 above)
Comment (Fenthick) [reply synthetics/test_conversions.go:225]: Yes, it is common idiom, when function returns error or `ok` flag as second argument. I don't use it everywhere, but in such case IMO it still looks readable and makes the code more compact.

File: synthetics/test_schema.go (hunk 2 above)
Comment (Fenthick) [reply synthetics/test_schema.go:12]: I'll change that

File: synthetics/test_schema.go (hunk 3 above)
Comment (Fenthick) [reply synthetics/test_schema.go:408]: I'll change that

File: examples/resources/kentik-synthetics_test/resource.tf [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    port     = 443
+    protocol = "icmp"
+    family   = "IP_FAMILY_V6"
+    servers = tolist([
+      "server-one",
+      "server-two",
+    ])
+    // Notice: specific values expected
+    // TODO(dfurman): provide allowed values
+    // target_type  = ""
```
Comment (mateuszmidor) [inline examples/resources/kentik-synthetics_test/resource.tf:75]: If I remember correctly, these two were read-only - server would reject the request carrying these params set

Comment (Fenthick) [reply examples/resources/kentik-synthetics_test/resource.tf:75]: Server responds as following:

> error: 400 Bad Request, status: 400 Bad Request, body: {{"code":3,"message":"At path: request.test.settings.target_type -- Expected a value of type `never`, but received: `\"foo\"`","details":[]}}


It look like client needs to pass specific value. @mmac-m3a Could you find out what are allowed values?

Comment (mmac-m3a) [reply examples/resources/kentik-synthetics_test/resource.tf:75]: I think Mateusz is right. The error message confirms it. It says it expect value of "type never", which is it's way to say that it is not expecting the attribute to be preset. I'll verify with Kentik engineering. Daniel, can you give me the complete request?

Comment (Fenthick) [reply examples/resources/kentik-synthetics_test/resource.tf:75]: This is the request that triggers the 400 response:
```bash
curl --location --request POST 'https://synthetics.api.kentik.com/synthetics/v202101beta1/tests' \
--header 'Content-Type: application/json' \
--header 'X-CH-Auth-Email: <email>' \
--header 'X-CH-Auth-API-Token: <token>' \
--data-raw '{
    "test": {
        "deviceId": "75702",
        "name": "example-hostname-test",
        "settings": {
            "agentIds": [
                "817",
                "818",
                "819"
            ],
            "count": 2,
            "expiry": 5001,
            "family": "IP_FAMILY_V6",
            "healthSettings": {
                "dnsValidCodes": [
                    1,
                    2,
                    3
                ],
                "httpLatencyCritical": 7,
                "httpLatencyWarning": 8,
                "httpValidCodes": [
                    200,
                    201
                ],
                "jitterCritical": 5,
                "jitterWarning": 6,
                "latencyCritical": 1,
                "latencyWarning": 2,
                "packetLossCritical": 3,
                "packetLossWarning": 4
            },
            "hostname": {
                "target": "www.example.com"
            },
            "limit": 4,
            "monitoringSettings": {
                "activationGracePeriod": "2",
                "activationTimeUnit": "m",
                "activationTimeWindow": "5",
                "activationTimes": "3",
                "notificationChannels": []
            },
            "period": 61,
            "ping": {
                "count": 5,
                "expiry": 3000,
                "period": 60
            },
            "port": 443,
            "protocol": "icmp",
            "reciprocal": false,
            "rollupLevel": 1,
            "servers": [
                "server-one",
                "server-two"
            ],
            "targetType": "foo",
            "targetValue": "bar",
            "tasks": [
                "ping",
                "traceroute"
            ],
            "trace": {
                "count": 3,
                "expiry": 22500,
                "limit": 30,
                "period": 60,
                "port": 33434,
                "protocol": "udp"
            },
            "useLocalIp": true
        },
        "status": "TEST_STATUS_ACTIVE",
        "type": "hostname"
    }
}'
```

**Pull request merged**